### PR TITLE
feat(keyspace): compute masklist attack keyspace from .hcmask line sum (#231)

### PR DESCRIPTION
**Pull request by unclesp1d3r**
## Summary

Computes the keyspace for hashcat **masklist** (`.hcmask`) attacks so they can be chunked for distribution (issue #231). Before this, a mode-3 attack referencing a `masklistId` (rather than an inline `mask`) got a `null` keyspace, so task generation could not split it and fell back to a single task.

`keyspace(masklist) = Σ per-line calculateMaskKeyspace(line)`, persisted to a new `mask_lists.keyspace` column and fanned out to dependent mode-3 attacks. Scoped to hashcat. Inline custom charsets, increment mode, and PRINCE remain `null` (single-task fallback) by design.

## Approach

- **Pure sum core** (`services/keyspace.ts`): `sumMasklistKeyspace` classifies each `.hcmask` line (blank/comment skipped; over-length, unescaped-comma custom-charset defs, `?1`-`?4` refs, and unknown tokens are uncomputable) and sums the per-line mask keyspace. A shared accumulator backs both a sync entry point (in-memory direct upload) and a streaming `sumMasklistKeyspaceFromStream` (worker/backfill) so the two paths cannot drift and a large file never buffers.
- **Whole-list-null rule:** any uncomputable line nulls the entire list — we under-chunk (single task) rather than mis-chunk on a wrong total.
- **Storage column:** `mask_lists.keyspace varchar(255)` (migration `0019`); always-string, mirroring `attacks.keyspace`.
- **Fan-out:** `computeAndPersistMasklistKeyspace` (shared by the line-count worker and a one-shot backfill) streams → sums → persists → recomputes dependent attack keyspaces. Masklists fan out **unconditionally**, including on a `null` result, so a re-upload to an uncomputable file propagates the null instead of leaving dependents stale.
- **Backfill:** `scripts/backfill-masklist-keyspace.ts` sizes pre-existing masklists; resilient per-row (one unreadable file logs and continues; the run exits non-zero if any row failed).

## hcmask grammar handled

- `?l ?u ?d ?s ?a ?h ?H ?b` charset sizes; literals contribute 1; `??` is a literal `?`.
- `#` comments skipped; `\#` is a real mask line.
- Unescaped comma = inline custom-charset definition → uncomputable (e.g. `?d?l,abc` is keyspace 1, not 260). The guard uses **backslash-run parity** matching hashcat's escape handling: `\,` and `\\\,` are escaped (literal comma), `\\,` is a real separator. A naive single-preceding-char check mis-classifies `\\,`.

## Tests

- `keyspace.test.ts` — line classification, summation, bigint-string precision, the escape-parity edge cases (`\\,` → null, `\\\,` → computable), and the streaming twin.
- `workers/line-count.test.ts` — masklist sum → persist → fan-out, uncomputable → null, storage-failure-no-partial-write.
- `services/resources-upload.test.ts` (new, isolated phase) — direct-upload sizing + the unconditional fan-out on `null` (the asymmetry fix).
- `services/attacks-complexity.test.ts`, `services/line-count-trigger.test.ts` — masklist lookup, enqueue, and fan-out wiring.

`just ci-check` green locally (backend: 742 pass / 0 fail; Playwright e2e: 11 passed).

## Scope / non-goals

Inline custom charsets, increment mode, and PRINCE stay `null` (single-task fallback). No new UI surface — keyspace is an internal task-distribution input.

## Residual review findings

One P3 reliability follow-up is recorded in `docs/residual-review-findings/feat-masklist-keyspace.md`: an uncomputable masklist's keyspace stays `null` forever, so the uncounted-resource sweep re-enqueues a (no-op) line-count job on each campaign trigger. Bounded and low-impact; the clean fix needs a `keyspace_state` design that is out of scope for #231.

## Notes

Closes #231. AI tooling (Claude Code) assisted with implementation and review; every line was reviewed and is owned by the author.

### CODE REVIEW

File: packages/backend/src/services/resources.ts [hunk 1]
```diff
@@ -588,6 +595,7 @@ export async function createResource(
 function lineCountTypeForResourceType(resourceType: string): LineCountResourceType | null {
   if (resourceType === 'wordlists') return 'wordlist'
   if (resourceType === 'rulelists') return 'rulelist'
+  if (resourceType === 'masklists') return 'masklist'
   return null
```
[the remark this answers is not in the document]
Comment (unclesp1d3r) [reply packages/backend/src/services/resources.ts:600]: > the doc comment immediately above still says masklists are not a keyspace input

Addressed in b4edecc: updated the `lineCountTypeForResourceType` doc comment. Masklists now map to `'masklist'` and are worker-sized via their summed mask keyspace (#231), so the comment no longer implies they are excluded.

File: packages/backend/src/services/attacks/complexity.ts [hunk 2]
```diff
@@ -62,6 +63,17 @@ export async function loadKeyspaceInputs(
     const cfg = attack.advancedConfiguration as Record<string, unknown>
     if (typeof cfg['mask'] === 'string') inputs.mask = cfg['mask']
   }
+  // A masklist-backed mode-3 attack (no inline mask) reads the masklist's
+  // precomputed summed keyspace. Null when uncomputable or not yet counted.
+  if (inputs.mask === undefined && attack.masklistId !== null) {
```
[the remark this answers is not in the document]
Comment (unclesp1d3r) [reply packages/backend/src/services/attacks/complexity.ts]: > Guard this lookup to mode 3 to avoid unnecessary DB queries (and avoid implying other modes support masklists).

Addressed in b4edecc: the `mask_lists.keyspace` lookup in `loadKeyspaceInputs` is now gated on `attack.mode === 3`. Since `masklistKeyspace` is a mode-3-only input (`calculateAttackKeyspace` ignores it for every other mode), this skips the wasted query for a stray `masklistId` on a non-mask attack and makes the mode-3 intent explicit.

File: packages/backend/tests/unit/services/resources-upload.test.ts [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+const IS_ISOLATED = process.env['RESOURCES_UPLOAD_TEST_ISOLATED'] === '1'
+
+if (!IS_ISOLATED) {
+  describe('resources-upload (skipped — runs in isolated phase)', () => {
+    test('runs only with RESOURCES_UPLOAD_TEST_ISOLATED=1', () => {
+      // oxlint-disable-next-line no-console -- surface phase-gating drift in CI logs
+      console.warn(
+        '[resources-upload] skipped — set RESOURCES_UPLOAD_TEST_ISOLATED=1 to run; this suite mocks db so it must NOT run in the shared phase.'
+      )
+      // Fail loud if a CI misconfig drops the isolated phase.
```
[the remark this answers is not in the document]
Comment (unclesp1d3r) [reply packages/backend/tests/unit/services/resources-upload.test.ts]: > the assertion only checks the env var is unset (which would still pass even if the isolated-phase run is removed). Reword the comment

Addressed in b4edecc: reworded the stub comment to describe what is actually asserted — that the gate is unset in this (shared) phase, i.e. the DB-mocking suite is being skipped here rather than leaking its mocks process-wide. The `console.warn` above is the part that surfaces a dropped isolated phase in CI logs.

File: packages/shared/src/db/schema.ts [hunk 4]
```diff
@@ -388,6 +388,10 @@ export const maskLists = pgTable('mask_lists', {
   fileRef: jsonb('file_ref').default({}),
   lineCount: bigint('line_count', { mode: 'number' }),
   fileSize: bigint('file_size', { mode: 'number' }),
+  // Summed keyspace of the masklist (Σ per-line calculateMaskKeyspace), a
+  // decimal string mirroring attacks.keyspace. Null when uncomputable
+  // (custom-charset / unknown-token lines) or not yet counted (#231).
+  keyspace: varchar('keyspace', { length: 255 }),
```
[the remark this answers is not in the document]
Comment (unclesp1d3r) [reply packages/shared/src/db/schema.ts:394]: > Separate pending vs permanently-uncomputable state for `mask_lists.keyspace` ... Use an explicit computation state (e.g., `keyspace_state`)

Acknowledged and tracked as a follow-up rather than fixed in this PR — this is the same `keyspace_state` design you tagged "Heavy lift". It requires a schema change (new enum column + backfill) that is out of scope for #231. It is recorded as a tracked residual in `docs/residual-review-findings/feat-masklist-keyspace.md` (P3), including the exact `pending | computed | uncomputable` enum and the sweep-predicate change you suggest.

The current `null`-overload is correct (keyspace results are never wrong; an uncomputable masklist degrades to the single-task fallback). The only cost is bounded, jobId-deduped re-enqueue work on each campaign trigger — the doc captures that rationale in full. Resolving here since the concern is durably recorded; the follow-up will reference it.
